File: google-cloud-bigquery/src/test/java/docs/javadsl/GoogleBigQuerySourceCustomParserDoc.java
/*
 * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.javadsl;

// #custom-parser-imports
// Imports
import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.http.javadsl.unmarshalling.Unmarshaller;
import akka.stream.Materializer;
import akka.stream.alpakka.googlecloud.bigquery.BigQueryConfig;
import akka.stream.alpakka.googlecloud.bigquery.BigQueryJsonProtocol;
import akka.stream.alpakka.googlecloud.bigquery.javadsl.BigQueryCallbacks;
import akka.stream.alpakka.googlecloud.bigquery.javadsl.GoogleBigQuerySource;
import akka.stream.javadsl.Source;
import akka.util.ByteString;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
// End imports

// #custom-parser-imports

public class GoogleBigQuerySourceCustomParserDoc {

  // #custom-parser
  static class User {
    String uid;
    String name;

    User(String uid, String name) {
      this.uid = uid;
      this.name = name;
    }
  }

  static ObjectMapper objectMapper = new ObjectMapper();

  static Unmarshaller<ByteString, JsonNode> jsonUnmarshaller =
      Unmarshaller.sync(
          byteString -> {
            try {
              return objectMapper.readTree(byteString.toArray());
            } catch (IOException ex) {
              throw new RuntimeException(ex);
            }
          });

  static Unmarshaller<JsonNode, BigQueryJsonProtocol.Response> responseUnmarshaller =
      Unmarshaller.sync(
          jsonNode -> {
            Optional<BigQueryJsonProtocol.JobReference> jobReference =
                Optional.ofNullable(jsonNode.path("jobReference"))
                    .map(
                        jobReferenceJsonNode -> {
                          Optional<String> jobId =
                              Optional.ofNullable(jobReferenceJsonNode.path("jobId"))
                                  .map(JsonNode::textValue);
                          return BigQueryJsonProtocol.createJobReference(jobId);
                        });

            Optional<String> pageToken =
                Optional.ofNullable(jsonNode.get("pageToken")).map(JsonNode::textValue);
            Optional<String> nextPageToken =
                Optional.ofNullable(jsonNode.get("nextPageToken")).map(JsonNode::textValue);

            Optional<Boolean> jobComplete =
                Optional.ofNullable(jsonNode.get("jobComplete")).map(JsonNode::booleanValue);

            return BigQueryJsonProtocol.createResponse(
                jobReference, pageToken, nextPageToken, jobComplete);
          });

  static Unmarshaller<JsonNode, BigQueryJsonProtocol.ResponseRows<User>> rowsUnmarshaller =
      Unmarshaller.sync(
          jsonNode -> {
            Optional<List<User>> rows =
                Optional.ofNullable(jsonNode.get("rows"))
                    .map(
                        rowsJsonNode -> {
                          List<User> users = new ArrayList(rowsJsonNode.size());
                          for (int i = 0; i < rowsJsonNode.size(); ++i) {
                            JsonNode userJsonNode = rowsJsonNode.get(i);
                            users.add(
                                new User(
                                    userJsonNode.get("uid").textValue(),
                                    userJsonNode.get("name").textValue()));
                          }

                          return users;
                        });

            return BigQueryJsonProtocol.createResponseRows(rows);
          });

  private static Source<User, NotUsed> customParser() {
    ActorSystem system = ActorSystem.create();
    Materializer materializer = Materializer.createMaterializer(system);
    BigQueryConfig config =
        BigQueryConfig.create(
            "project@test.test",
            "privateKeyFromGoogle",
            "projectID",
            "bigQueryDatasetName",
            system);
    return GoogleBigQuerySource.runQuery(
        "SELECT uid, name FROM bigQueryDatasetName.myTable",
        rowsUnmarshaller,
        responseUnmarshaller,
        jsonUnmarshaller,
        BigQueryCallbacks.ignore(),
        config);
  }
  // #custom-parser

}

File: google-cloud-bigquery/src/test/scala/docs/scaladsl/GoogleBigQuerySourceCustomParserDoc.scala
/*
 * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.scaladsl

import akka.NotUsed
import akka.actor.ActorSystem
import akka.stream.alpakka.googlecloud.bigquery.BigQueryConfig
import akka.stream.alpakka.googlecloud.bigquery.scaladsl.{BigQueryCallbacks, GoogleBigQuerySource}
import akka.stream.scaladsl.Source

import scala.concurrent.Future

// Separate class to circumvent ambiguous implicits for unmarshalling
class GoogleBigQuerySourceCustomParserDoc {

  import akka.stream.alpakka.googlecloud.bigquery.client.ResponseJsonProtocol

  implicit val system = ActorSystem()
  val config = BigQueryConfig("project@test.test", "privateKeyFromGoogle", "projectID", "bigQueryDatasetName")

  //#custom-parser
  import akka.http.scaladsl.unmarshalling.{FromByteStringUnmarshaller, Unmarshaller}
  import io.circe.{jawn, Decoder, Json}
  import io.circe.generic.semiauto

  implicit val jsonUnmarshaller: FromByteStringUnmarshaller[Json] = Unmarshaller { _ => byteString =>
    Future.fromTry(jawn.parseByteBuffer(byteString.asByteBuffer).toTry)
  }

  implicit def fromJsonUnmarshaller[T](implicit decoder: Decoder[T]): Unmarshaller[Json, T] = Unmarshaller {
    _ => json =>
      Future.fromTry(decoder.decodeJson(json).toTry)
  }

  case class User(uid: String, name: String)

  implicit val userDecoder: Decoder[User] = Decoder.instance { c =>
    val f = c.downField("f")
    for {
      uid <- f.downN(0).downField("v").as[String]
      name <- f.downN(1).downField("v").as[String]
    } yield User(uid, name)
  }

  implicit val jobReferenceDecoder: Decoder[ResponseJsonProtocol.JobReference] = semiauto.deriveDecoder
  implicit val responseDecoder: Decoder[ResponseJsonProtocol.Response] = semiauto.deriveDecoder

  implicit def rowsDecoder[T](implicit decoder: Decoder[T]): Decoder[ResponseJsonProtocol.ResponseRows[T]] =
    semiauto.deriveDecoder

  val userStream: Source[User, NotUsed] =
    GoogleBigQuerySource[User]
      .runQuery("SELECT uid, name FROM bigQueryDatasetName.myTable", BigQueryCallbacks.ignore, config)
  //#custom-parser
}
